### PR TITLE
jv-spring-boot-web

### PR DESCRIPTION
**Pull request by qreqit**

### CODE REVIEW

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package mate.academy.springbootwebgreqit.model;
+
+import jakarta.persistence.*;
+import lombok.Data;
+
+import java.math.BigDecimal;
+
+@Data
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:8]: @Data on JPA entities can have unexpected effect, use @Getter/@Setter instead

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
+
+    public String getAuthor() {
+        return author;
+    }
+
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:41]: remove boilerplate, you have lombok

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String title;
+    private String author;
+    private String isbn;
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:20]: add column constraints  (nullability, length etc)

File: src/main/java/mate/academy/springbootwebgreqit/config/MapperComfig.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import org.mapstruct.NullValueCheckStrategy;
+
+@org.mapstruct.MapperConfig(
+        componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl"
+)
+public class MapperComfig {
+}
```
Comment (okuzan) [inline src/main/java/mate/academy/springbootwebgreqit/config/MapperComfig.java:13]: add an empty line

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/springbootwebgreqit/dto/BookDto.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+@Setter
+public class BookDto {
+    private Long id;
+    private String title;
+    private String author;
+    private String isbn;
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
+
```
Comment (sarakhmen) [inline src/main/java/mate/academy/springbootwebgreqit/dto/BookDto.java:18]:
```suggestion
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java (hunk 3 above)
Comment (sarakhmen) [reply src/main/java/mate/academy/springbootwebgreqit/model/Book.java]: still not fixed

**Review by sarakhmen — APPROVED** [review]

File: pom.xml [hunk 6]
```diff
@@ -1,67 +1,122 @@
...
-		</dependency>
-		<dependency>
-			<groupId>org.springframework.boot</groupId>
-			<artifactId>spring-boot-starter-test</artifactId>
-			<scope>test</scope>
-		</dependency>
-	</dependencies>
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-core</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml:42]: it's better to use spring-boot-starter-data-jpa

File: pom.xml [hunk 7]
```diff
@@ -1,67 +1,122 @@
...
-						</exclude>
-					</excludes>
-				</configuration>
-			</plugin>
-		</plugins>
-	</build>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>4.28.0</version>
```
Comment (Elena-Bruyako) [inline pom.xml:50]: spring parent provide this version

File: pom.xml [hunk 8]
```diff
@@ -1,67 +1,122 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-maven-plugin</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml:55]: the same

File: pom.xml [hunk 9]
```diff
@@ -1,67 +1,122 @@
...
+
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-annotations</artifactId>
+            <version>3.5.6-Final</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework.data</groupId>
+            <artifactId>spring-data-jpa</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml:67]: remove

File: src/main/java/mate/academy/springbootwebgreqit/config/MapperComfig.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+import org.mapstruct.InjectionStrategy;
+import org.mapstruct.NullValueCheckStrategy;
+
+@org.mapstruct.MapperConfig(
+        componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl"
+)
+public class MapperComfig {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/config/MapperComfig.java:12]:
```suggestion
public class MapperConfig {
```

File: src/main/java/mate/academy/springbootwebgreqit/controller/BookController.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import mate.academy.springbootwebgreqit.dto.CreateBookRequestDto;
+import mate.academy.springbootwebgreqit.service.BookService;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value =  "/books")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/BookController.java:14]:
```suggestion
@RequestMapping("/books")
```

File: src/main/java/mate/academy/springbootwebgreqit/controller/BookController.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        return bookService.save(requestDto);
+    }
+
+    @PutMapping("/{id}")
+    public BookDto updateBook(@PathVariable Long id) {
+        return bookService.update(id);
+    }
+
+    @DeleteMapping("/{id}")
+    public void  deleteBook(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/BookController.java:39]:
```suggestion
    public void deleteBook(@PathVariable Long id) {
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+@Table(name = "books")
+public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, length = 100)
+    private String title;
+    @Column(nullable = false, length = 50)
+    private String author;
+    @Column(unique = true, length = 13)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:24]: nullable?

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    private String isbn;
+    @Column(nullable = false, precision = 10, scale = 2)
+    private BigDecimal price;
+    @Column(length = 2000)
+    private String description;
+    @Column(name = "cover_image", length = 255)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:30]:
```suggestion
    @Column(length = 255)
```

File: src/main/java/mate/academy/springbootwebgreqit/repository/BookRepository.java [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.springbootwebgreqit.repository;
+
+import mate.academy.springbootwebgreqit.dto.BookDto;
+import mate.academy.springbootwebgreqit.model.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface BookRepository extends JpaRepository<Book, Long> {
+    BookDto updateBookById (Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/repository/BookRepository.java:8]:
```suggestion
```

File: src/main/java/mate/academy/springbootwebgreqit/service/BookService.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+import mate.academy.springbootwebgreqit.dto.CreateBookRequestDto;
+
+import java.util.List;
+
+public interface BookService {
+    BookDto save(CreateBookRequestDto requestDto);
+
+    List<BookDto> findAll();
+
+    public BookDto findById(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/BookService.java:14]:
```suggestion
    BookDto findById(Long id);
```

File: src/main/java/mate/academy/springbootwebgreqit/service/BookService.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+    public void deleteById(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/BookService.java:16]:
```suggestion
    deleteById(Long id);
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/BookServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+        Book book = bookMapper.toModel(requestDto);
+        Book savedBook = bookRepository.save(book);
+        return bookMapper.toDto(savedBook);
+    }
+
+    @Override
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/BookServiceImpl.java:34]:
```suggestion
                .toList();
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/BookServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public BookDto findById(Long id) {
+        Book book = bookRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("Can't find biik by id:" + id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/BookServiceImpl.java:39]:
```suggestion
        Book book = bookRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("Can't find book by id:" + id));
```

File: src/main/java/mate/academy/springbootwebgreqit/service/impl/BookServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        bookRepository.deleteById(id);
+    }
+
+    @Override
+    public BookDto update(Long id) {
+        return bookRepository.updateBookById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/service/impl/BookServiceImpl.java:50]: implement logic

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
let's fix some comments

File: src/main/java/mate/academy/springbootwebgreqit/dto/UpdateBookRequestDto.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.springbootwebgreqit.dto;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class UpdateBookRequestDto {
+    private String title;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/dto/UpdateBookRequestDto.java:9]: see the task description of Update book

File: src/main/java/mate/academy/springbootwebgreqit/controller/BookController.java [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @PostMapping
+    public BookDto createBook(@RequestBody CreateBookRequestDto requestDto) {
+        return bookService.save(requestDto);
+    }
+
+    @PutMapping("/{id}")
+    public BookDto updateBook(@PathVariable Long id, UpdateBookRequestDto updateBookRequestDto) {
+        return bookService.update(id,updateBookRequestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/BookController.java:36]: In this PR you should only add update and delete endpoint


Book Endpoints:

GET: /api/books (Retrieve book catalog) (should be done in the previous PR)
GET: /api/books/{id} (Retrieve book details) (should be done in the previous PR)
POST: /api/books (Create a new book) (should be done in the previous PR)
PUT: /api/books/{id} (Update a specific book)

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java (hunk 13 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springbootwebgreqit/model/Book.java:24]: not fixed

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE books SET is_deleted = true where id = ?")
+@Where(clause = "is_deleted = false")
+@Table(name = "books")
+public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(length = 100)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:20]: nullable?

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    private String title;
+    @Column(length = 50)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:22]: nullable?

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    private String author;
+    @Column(unique = true, length = 13)
+    private String isbn;
+    @Column(precision = 10, scale = 2)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:26]: Add constraint for all required fields


FYI This should be done in previous HWs

File: src/main/resources/db/chandelog/changes/02-add-column-isDeleted-to-books-table.yaml [hunk 26]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-column-isDeleted-to-books-table
+      author: greqit
+      changes:
+        - addColumn:
+            tableName: books
+            columns:
+              - column:
+                  name: is_deleted
```
Comment (Elena-Bruyako) [inline src/main/resources/db/chandelog/changes/02-add-column-isDeleted-to-books-table.yaml:10]: replace to 01 changelog

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
PR includes many changes, that should be done in previous HWs

File: src/main/java/mate/academy/springbootwebgreqit/dto/UpdateBookRequestDto.java [hunk 27]
```diff
@@ -0,0 +1,18 @@
+package mate.academy.springbootwebgreqit.dto;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import java.math.BigDecimal;
+
+@Getter
+@Setter
+public class UpdateBookRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/dto/UpdateBookRequestDto.java:10]: What the difference in fields between UpdateBookRequestDto and CreateBookRequestDto?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request closed without merging**